pkg/catalog/manager/sync.go

```diff
@@ -2,12 +2,90 @@ package manager
 
 import (
 	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/client/management/v3"
 	"github.com/sirupsen/logrus"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
 )
 
+func (m *Manager) PrjCatalogSync(key string, obj *v3.ProjectCatalog) error {
+	m.commonSync(key, &obj.Catalog)
+	cmt := &CatalogManagerType{
+		catalog:        &obj.Catalog,
+		projectCatalog: obj,
+		catalogType:    client.ProjectCatalogType,
+	}
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	projectCatalog, err := m.projectCatalogClient.GetNamespaced(ns, name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	repoPath, commit, err := m.prepareRepoPath(obj.Catalog)
+	if err != nil {
+		v3.CatalogConditionRefreshed.False(projectCatalog)
+		v3.CatalogConditionRefreshed.ReasonAndMessageFromError(projectCatalog, err)
+		m.projectCatalogClient.Update(projectCatalog)
+		return err
+	}
+
+	if commit == projectCatalog.Status.Commit {
+		logrus.Debugf("Catalog %s is already up to date", projectCatalog.Name)
+		if v3.CatalogConditionRefreshed.IsUnknown(projectCatalog) {
+			v3.CatalogConditionRefreshed.True(projectCatalog)
+			v3.CatalogConditionRefreshed.Reason(projectCatalog, "")
+			m.projectCatalogClient.Update(projectCatalog)
+		}
+		return nil
+	}
```

Comment:
Is it possible to take lines 30-46 out into a common method that takes in "catalog *v3.Catalog" and updates the v3.CatalogConditionRefreshed and returns an err if any or nil. In the caller method, you can then do the "m.projectCatalogClient.Update" and return the err/nil

Reply:
v3.CatalogConditionRefreshed updates v3.Catalog in sync, and v3.ProjectCatalog in PrjCatalogSync methods, and then we're calling updates using different clients for each, so I think it's okay for them to be in separate methods

Reply:
I feel like there is too much common logic still. We have to think about extracting the common logic, and may be passing the update function as a callback....